server/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java

```diff
@@ -533,9 +538,21 @@ assert getByAllocationId(failedShard.shardId(), failedShard.allocationId().getId
                         // re-resolve replica as earlier iteration could have changed source/target of replica relocation
                         ShardRouting replicaShard = getByAllocationId(routing.shardId(), routing.allocationId().getId());
                         assert replicaShard != null : "failed to re-resolve " + routing + " when failing replicas";
-                        UnassignedInfo primaryFailedUnassignedInfo = new UnassignedInfo(UnassignedInfo.Reason.PRIMARY_FAILED,
-                            "primary failed while replica initializing", null, 0, unassignedInfo.getUnassignedTimeInNanos(),
-                            unassignedInfo.getUnassignedTimeInMillis(), false, AllocationStatus.NO_ATTEMPT, Collections.emptySet());
+                        boolean nodeIsRestarting = Optional.ofNullable(nodeShutdowns.get(replicaShard.currentNodeId()))
+                            .map(shutdownInfo -> shutdownInfo.getType().equals(SingleNodeShutdownMetadata.Type.RESTART))
+                            .orElse(false);
+                        UnassignedInfo primaryFailedUnassignedInfo = new UnassignedInfo(
+                            UnassignedInfo.Reason.PRIMARY_FAILED,
+                            "primary failed while replica initializing",
+                            null,
+                            0,
+                            unassignedInfo.getUnassignedTimeInNanos(),
+                            unassignedInfo.getUnassignedTimeInMillis(),
+                            false,
```

Comment:
It is a bit annoying that we have to calculate `nodeIsRestarting` when we pass in `false` to delayed anyway. I wonder if we were better off by making delayed an enum with values "NORMAL", "SHUTDOWN" (and possibly use "NO" instead of null)? That would avoid calculating `nodeIsRestarting` above for no good reason.

Reply:
This no longer needs to be calculated here, as the reason is `PRIMARY_FAILED`.